src/Temporalio/Workflows/WorkflowDefinition.cs

```diff
@@ -265,35 +282,41 @@ public static WorkflowDefinition CreateWithoutAttribute(string name, Type type)
                 name: name,
                 type: type,
                 runMethod: runMethod!,
-                instantiable: instantiable,
-                initConstructor: initConstructor,
+                creator: creator,
                 signals: signals,
                 queries: queries);
         }
 
-        private static WorkflowDefinition CreateFromType(Type type)
+        /// <summary>
+        /// Get the workflow definition for the type the given run method is on. Expects the method
+        /// to have <see cref="WorkflowRunAttribute" />.
+        /// </summary>
+        /// <param name="runMethod">Workflow run method.</param>
+        /// <returns>Workflow definition for the type the run method is on.</returns>
+        public static WorkflowDefinition FromRunMethod(MethodInfo runMethod)
         {
-            // Unwrap the type
-            type = Refs.GetUnderlyingType(type);
-
-            // Get the main attribute, but throw immediately if it is not present
-            var attr = type.GetCustomAttribute<WorkflowAttribute>(false) ??
-                throw new ArgumentException($"{type} missing Workflow attribute");
-
-            // Use given name or default
-            var name = attr.Name;
-            if (name == null)
+            if (runMethod.GetCustomAttribute<WorkflowRunAttribute>() == null)
             {
-                name = type.Name;
-                // If type is an interface and name has a leading I followed by another capital, trim it
-                // off
-                if (type.IsInterface && name.Length > 1 && name[0] == 'I' && char.IsUpper(name[1]))
-                {
-                    name = name.Substring(1);
-                }
+                throw new ArgumentException($"{runMethod} missing WorkflowRun attribute");
             }
+            // We intentionally use reflected type because we don't allow inheritance of run methods
+            // in any way, they must be explicitly defined on the type
+            return Create(runMethod.ReflectedType ??
+                throw new ArgumentException($"{runMethod} has no reflected type"));
```

Comment:
Would be nice to have the error message make reference to how to fix this

Reply:
It's a very rare case, see https://learn.microsoft.com/en-us/dotnet/api/system.reflection.memberinfo.reflectedtype?view=net-7.0#remarks. It's only on module-level methods which you can't even write in C# (but the .NET IL technically supports it).